src/doc/man/fedfs-lookup-junction.8
.\"@(#)fedfs-lookup-junction.8"
.\"
.\" @file doc/man/fedfs-lookup-junction.8
.\" @brief man page for fedfs-lookup-junction client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-LOOKUP-JUNCTION 8 "2 July 2013"
.SH NAME
fedfs-lookup-junction \- send a FEDFS_LOOKUP_JUNCTION ADMIN protocol request
.SH SYNOPSIS
.B fedfs-lookup-junction
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.RB [ \-t
.IR resolvetype ]
.I path
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-lookup-junction (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_LOOKUP_JUNCTION request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_LOOKUP_JUNCTION request causes a remote server
to reveal the contents of a junction,
or to report cached or immediate NSDB lookup results as that server sees them.
The contents of a FedFS junction are an FSN UUID and an NSDB name and port.
.P
The
.BR fedfs-lookup-junction (8)
command takes a single positional parameter which is the
pathname on the remote server of the junction to be looked up.
The pathname is relative to the root
of the local file system on the remote server.
.P
Resolving a junction means performing an NSDB query with the contents
of the junction to obtain a list of fileset locations, or FSLs,
matching the stored FSN UUID.
The meaning of these is described in more detail in
.BR fedfs (7).
.P
There are three distinct types of junction lookup:
.TP
.B none
The remote server reports the actual contents of the junction stored
on its local disk.
This includes an FSN UUID and the name and port of an NSDB.
If the
.B \-t
option is not specified, this type of lookup is performed.
.TP
.B cache
The remote server reports lookup results it may have cached from
previous junction lookup requests.
This includes an FSN UUID, the name and port of an NSDB, and the cached
list of fileset locations matching the FSN UUID in the junction.
Not all FedFS ADMIN service implementations support this type of request.
.TP
.B nsdb
The remote server performs a fresh junction lookup
and the results are returned.
This includes an FSN UUID, the name and port of an NSDB,
and a list of fileset locations matching the FSN UUID in the junction.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-lookup-junction (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.IP "\fB\-p, \-\-path=\fIpathname\fP"
Specifies the location on the remote server where the target FedFS junction
resides.
This pathname is relative to the remote server's physical root directory,
not the remote server's NFS pseudoroot.
.IP "\fB\-t, \-\-resolvetype=\fItype\fP"
Specifies the desired type of resolution.  Valid values for
.I type
are
.BR 0 ,
.BR none ,
.BR fedfs_resolve_none ,
.BR 1 ,
.BR cache ,
.BR fedfs_resolve_cache ,
.BR 2 ,
.BR nsdb ", or"
.BR fedfs_resolve_nsdb .
The value is not case-sensitive.
If this option is not specified, the default value is
.BR none .
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The pathname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADNAME
A component of the pathname consisted of valid UTF-8 characters
supported by the remote server,
but the name is not supported by the remote server
as a valid name for the current operation.
.TP
.B FEDFS_ERR_NAMETOOLONG
The length of the pathname exceeds the remote server’s implementation limit.
.TP
.B FEDFS_ERR_LOOP
Too many symbolic links were encountered in resolving the pathname.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_NOTJUNCT
The pathname does not end in a junction.
.TP
.B FEDFS_ERR_NOTLOCAL
The pathname contains a junction in any position other than the last component.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The remote server was unable to find a route to the NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The remote server determined that the NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The remote server was unable to establish a connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The remote server was unable to authenticate
and establish a secure connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
An LDAP error occurred on the connection between the remote server and NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP,
and allows an LDAP protocol error value to be returned to the client.
.TP
.B FEDFS_ERR_NSDB_NONCE
The remote server was unable to locate the NCE in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSN
The remote server was unable to locate the given FSN in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSL
The remote server was unable to locate any FSLs for the given FSN
in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The remote server received a malformed response from the NSDB.
This includes situations when an NSDB entry (e.g. FSN or FSL)
is missing a required attribute.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The remote server does not have any connection
parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The remote server received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP_REFERRAL,
and allows an LDAP protocol error value to be returned back to the client.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The remote server received an LDAP referral that it chose not to follow,
either because the remote server does not support following LDAP referrals
or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The remote server received an LDAP referral that it chose not to follow
because the remote server had no NSDB parameters for the NSDB
targeted by the LDAP referral.
.TP
.B FEDFS_ERR_PATH_TYPE_UNSUPP
The remote server does not support the specified FedFsPathType value.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.TP
.B FEDFS_ERR_NO_CACHE
The remote server does not implement an FSN-to-FSL cache.
.TP
.B FEDFS_ERR_UNKNOWN_CACHE
The software receiving the ONC RPC request is unaware if the remote server
implements an FSN-to-FSL cache or unable to communicate with the
local FSN-to-FSL cache if it exists.
.TP
.B FEDFS_ERR_NO_CACHE_UPDATE
The remote server was unable to update its FSN-to-FSL cache.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that your domain's NSDB hostname is
.IR nsdb.example.net .
You have created a FedFS junction on remote server
.IR fs.example.net .
To see how the junction appears on the file server, use:
.RS
.sp
$ fedfs-lookup-junction -h fs.example.net /export/junction1
.br
Call completed successfully
.br
FSN UUID: 89c6d208-7280-11e0-9f1d-000c297fd679
.br
NSDB: nsdb.example.net:389
.sp
.RE
To see real-time junction lookup results as the remote server sees them, use:
.RS
.sp
$ fedfs-lookup-junction -h fs.example.net -t nsdb /export/junction1
.br
Server returned FEDFS_ERR_NSDB_NOFSN
.sp
.RE
In this example, the junction exists on the file server,
but the domain's NSDB has not yet been updated to contain a list of
fileset locations for the FSN UUID contained in the junction.
The file server is therefore not able to resolve the junction.
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/src/fedfsc/fedfs-lookup-junction.c
/**
 * @file src/fedfsc/fedfs-lookup-junction.c
 * @brief Send a FEDFS_LOOKUP_JUNCTION RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010, 2013 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <ldap.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char fedfs_lookup_junction_opts[] = "?dh:n:t:s:";

/**
 * Long form command line options
 */
static const struct option fedfs_lookup_junction_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nettype", 1, NULL, 'n', },
	{ "resolvetype", 1, NULL, 't', },
	{ "security", 1, NULL, 's', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 * @return program exit status
 */
static int
fedfs_lookup_junction_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"[-t <none|cache|nsdb>] path\n\n", progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fprintf(stderr, "\t-t, --resolvetype    Type of desired result (default: 'none')\n");
	fprintf(stderr, "\t-s, --security       RPC security level\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	return EXIT_FAILURE;
}

/**
 * Display one FSL in a FEDFS_LOOKUP_JUNCTION result
 *
 * @param fsl FSL record to display
 */
static void
fedfs_lookup_junction_print_fsl(struct admin_fsl *fsl)
{
	unsigned int i;

	printf(" FSL UUID: %s\n", fsl->al_uuid);

	printf(" FSL hostname: %s:%u\n",
		fsl->al_hostname, fsl->al_port);

	if (fsl->al_pathname[0] == NULL)
		printf(" FSL NFS pathname: /\n");
	else {
		printf(" FSL NFS pathname: ");
		for (i = 0; fsl->al_pathname[i] != NULL; i++)
			printf("/%s", fsl->al_pathname[i]);
		printf("\n");
	}
}

/**
 * Display FSLs returned by a FEDFS_LOOKUP_JUNCTION request
 *
 * @param fsl list of FSLs returned from the server
 */
static void
fedfs_lookup_junction_print_fsls(struct admin_fsl *fsl)
{
	if (fsl == NULL)
		return;

	printf("Returned FSLs:\n");
	while (fsl != NULL) {
		fedfs_lookup_junction_print_fsl(fsl);
		fsl = fsl->al_next;
	}
}

/**
 * Look up a junction on a remote fileserver
 *
 * @param host an initialized and opened admin_t
 * @param path_array an array of NUL-terminated C strings containing pathname components
 * @param request requested resolution type
 * @return program exit status
 */
static FedFsStatus
fedfs_lookup_junction_try(admin_t host, char * const *path_array, int request)
{
	struct admin_fsl *fsls = NULL;
	struct admin_fsn *fsn = NULL;
	int status, err;

	status = EXIT_FAILURE;
	switch (request) {
	case 0:
		err = admin_lookup_junction_none(host, path_array, &fsn);
		break;
	case 1:
		err = admin_lookup_junction_cached(host, path_array,
								&fsn, &fsls);
		break;
	case 2:
		err = admin_lookup_junction_nsdb(host, path_array,
								&fsn, &fsls);
		break;
	default:
		xlog(L_ERROR, "Unrecognized request");
		goto out;
	}

	switch (err) {
	case 0:
		break;
	case EACCES:
		xlog(L_ERROR, "%s: access denied", admin_hostname(host));
		xlog(D_GENERAL, "%s",
			admin_perror(host, admin_hostname(host)));
		goto out;
	case EIO:
		xlog(L_ERROR, "%s",
			admin_perror(host, admin_hostname(host)));
		goto out;
	default:
		xlog(L_ERROR, "ADMIN client: %s", strerror(err));
		goto out;
	}

	switch (admin_status(host)) {
	case FEDFS_OK:
		printf("FSN UUID: %s\n", fsn->af_uuid);
		printf("NSDB: %s:%u\n",
			fsn->af_nsdb.an_hostname,
			fsn->af_nsdb.an_port);
		if (request > 0)
			fedfs_lookup_junction_print_fsls(fsls);
		status = EXIT_SUCCESS;
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fprintf(stderr, "LDAP result code (%d): %s\n",
			admin_ldaperr(host),
			ldap_err2string(admin_ldaperr(host)));
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		printf("No connection parameters found\n");
		break;
	default:
		nsdb_print_fedfsstatus(admin_status(host));
	}
	admin_free_fsls(fsls);
	admin_free_fsn(fsn);

out:
	return status;
}

/**
 * Look up a junction on a remote fileserver
 *
 * @param hostname NUL-terminated UTF-8 string containing ADMIN server's hostname
 * @param nettype NUL-terminated C string containing nettype to use for connection
 * @param security NUL-terminated C string containing RPC security mode
 * @param path_array an array of NUL-terminated C strings containing pathname components
 * @param request requested resolution type
 * @return program exit status
 */
static int
fedfs_lookup_junction_host(const char *hostname, const char *nettype,
		const char *security, char * const *path_array, int request)
{
	admin_t host;
	int status;

	status = EXIT_FAILURE;
	switch (admin_create(hostname, nettype, security, &host)) {
	case 0:
		status = fedfs_lookup_junction_try(host, path_array, request);
		admin_release(host);
		break;
	case EINVAL:
		xlog(L_ERROR, "Invalid command line parameter");
		break;
	case EACCES:
		xlog(L_ERROR, "Failed to authenticate server");
		break;
	case EKEYEXPIRED:
		xlog(L_ERROR, "User credentials not found");
		break;
	default:
		xlog(L_ERROR, "%s",
			admin_open_perror(hostname));
	}

	return status;
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype, *path, *security, *resolvetype;
	int arg, status, request;
	FedFsStatus retval;
	char **path_array;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	hostname = "localhost";
	nettype = "netpath";
	security = "unix";
	resolvetype = "none";
	while ((arg = getopt_long(argc, argv, fedfs_lookup_junction_opts,
				fedfs_lookup_junction_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		case 's':
			security = optarg;
			break;
		case 't':
			resolvetype = optarg;
			break;
		default:
			fprintf(stderr, "Invalid command line argument: %c\n", (char)arg);
		case '?':
			return fedfs_lookup_junction_usage(progname);
		}
	}
	if (argc == optind + 1)
		path = argv[optind];
	else if (argc > optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		return fedfs_lookup_junction_usage(progname);
	} else {
		fprintf(stderr, "No junction pathname was specified\n");
		return fedfs_lookup_junction_usage(progname);
	}

	if (strcmp(resolvetype, "0") == 0 ||
	    strcasecmp(resolvetype, "none") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_none") == 0)
		request = 0;
	else if (strcmp(resolvetype, "1") == 0 ||
	    strcasecmp(resolvetype, "cache") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_cache") == 0)
		request = 1;
	else if (strcmp(resolvetype, "2") == 0 ||
	    strcasecmp(resolvetype, "nsdb") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_nsdb") == 0)
		request = 2;
	else {
		fprintf(stderr, "Unrecognized resolvetype\n");
		return EXIT_FAILURE;
	}

	retval = nsdb_posix_to_path_array(path, &path_array);
	if (retval != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(retval));
		return EXIT_FAILURE;
	}

	status = fedfs_lookup_junction_host(hostname, nettype,
						security, path_array, request);

	nsdb_free_string_array(path_array);
	return status;
}
